<a href="https://colab.research.google.com/github/guojiaren/Latia113-1/blob/main/Copy_of_Welcome_To_Colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import statsmodels.api as sm
import numpy as np

df2 = pd.DataFrame(np.random.randn(100, 4), columns=list("ABCD"))
X = df2['A']
y = df2['B']
# Add a constant to the model (for the intercept term)
X = sm.add_constant(X)
X
# Fit the regression model
model = sm.OLS(y, X).fit()

# Display the regression results
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                      B   R-squared:                       0.019
Model:                            OLS   Adj. R-squared:                  0.009
Method:                 Least Squares   F-statistic:                     1.897
Date:                Thu, 07 Nov 2024   Prob (F-statistic):              0.172
Time:                        09:14:07   Log-Likelihood:                -145.51
No. Observations:                 100   AIC:                             295.0
Df Residuals:                      98   BIC:                             300.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0425      0.105     -0.406      0.6

In [ ]:
import pandas as pd
import statsmodels.api as sm
import numpy as np

# 創建資料框
df2 = pd.DataFrame(np.random.randn(100, 4), columns=list("ABCD"))

# 定義因變數 y 和自變數 X（多重回歸使用多個自變數）
X = df2.iloc[:, 1:3]  # 取 B 和 C 兩列作為自變數
y = df2['A']  # 目標變數是 A

# 為 X 加上常數項（截距）
X = sm.add_constant(X)

# 擬合回歸模型
model = sm.OLS(y, X).fit()

# 顯示回歸結果
print(model.summary())


                            OLS Regression Results                            
Dep. Variable:                      A   R-squared:                       0.018
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.8960
Date:                Thu, 07 Nov 2024   Prob (F-statistic):              0.412
Time:                        09:19:13   Log-Likelihood:                -149.51
No. Observations:                 100   AIC:                             305.0
Df Residuals:                      97   BIC:                             312.8
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.1641      0.111     -1.478      0.1

Dep. Variable:A is - see B because B is answer (-) and Dep. Variable: A is + see C because of C is answer + . Prob (F-statistic) is 0.412 so this answer is false because P value is less than 0.1

In [1]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

df = pd.read_csv('A_stat_data_112.csv')
# 假設要預測 'correct_rate' 並使用其他數值型變量作為特徵
df = df.dropna(subset=['correct_rate']) # 去除 'correct_rate' 為 NaN 的行
X = df[['grade', 'is_completed', 'answer_duration_sec']] # 例選三個特徵
y = df['correct_rate']
# 分割訓練集和測試集
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# 訓練隨機森林回歸模型
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)
# 預測及評估
y_pred = rf_model.predict(X_test)
print("均方誤差 (MSE):", mean_squared_error(y_test, y_pred))

# 將特徵重要性與特徵名稱對應起來
feature_importance = pd.DataFrame({
'Feature': X.columns,
'Importance': rf_model.feature_importances_
})
# 顯示特徵重要性
print(feature_importance)

#R2
from sklearn.metrics import r2_score
r2 = r2_score(y_test, y_pred)
print("R^2:", r2)

均方誤差 (MSE): 0.07500122426622928
               Feature  Importance
0                grade    0.037625
1         is_completed    0.000000
2  answer_duration_sec    0.962375
R^2: 0.2585571385908416
